cocos/ui/UIWebView/UIWebViewImpl-win32.cpp

```diff
@@ -0,0 +1,1308 @@
+ in the Software without restriction, including without limitation the rights
+ to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ copies of the Software, and to permit persons to whom the Software is
+ furnished to do so, subject to the following conditions:
+ 
+ The above copyright notice and this permission notice shall be included in
+ all copies or substantial portions of the Software.
+ 
+ THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ THE SOFTWARE.
+ ****************************************************************************/
+
+#if CC_TARGET_PLATFORM == CC_PLATFORM_WIN32
+
+#include "UIWebViewImpl-win32.h"
+#include "UIWebView.h"
+#include "base/CCDirector.h"
+#include "platform/CCFileUtils.h"
+#include "platform/CCGLView.h"
+#include "base/base64.h"
+#include "ui/UIHelper.h"
+
+#define WIN32_LEAN_AND_MEAN
+#include <Shlwapi.h>
+#include <codecvt>
+#include <utility>
+#include <stdlib.h>
+#include <Windows.h>
+
+#pragma comment(lib, "user32.lib")
+#pragma comment(lib, "Shlwapi.lib")
+
+// Edge/Chromium headers and libs
+#include "WebView2.h"
+#pragma comment(lib, "ole32.lib")
+#pragma comment(lib, "oleaut32.lib")
+
+#include "platform/CCPlatformConfig.h"
+
+using msg_cb_t = std::function<void(const std::string)>;
+
+// Convert ASCII hex digit to a nibble (four bits, 0 - 15).
+//
+// Use unsigned to avoid signed overflow UB.
+static inline unsigned char hex2nibble(unsigned char c)
+{
+    if (c >= '0' && c <= '9')
+    {
+        return c - '0';
+    }
+    else if (c >= 'a' && c <= 'f')
+    {
+        return 10 + (c - 'a');
+    }
+    else if (c >= 'A' && c <= 'F')
+    {
+        return 10 + (c - 'A');
+    }
+    return 0;
+}
+
+// Convert ASCII hex string (two characters) to byte.
+//
+// E.g., "0B" => 0x0B, "af" => 0xAF.
+static inline char hex2char(const char* p)
+{
+    return hex2nibble(p[0]) * 16 + hex2nibble(p[1]);
+}
+
+inline std::string url_encode(const std::string s)
+{
+    std::string encoded;
+    for (unsigned int i = 0; i < s.length(); i++)
+    {
+        auto c = s[i];
+        if (isalnum(c) || c == '-' || c == '_' || c == '.' || c == '~')
+        {
+            encoded = encoded + c;
+        }
+        else
+        {
+            char hex[4];
+            snprintf(hex, sizeof(hex), "%%%02x", c);
+            encoded = encoded + hex;
+        }
+    }
+    return encoded;
+}
+
+inline std::string url_decode(const std::string st)
+{
+    std::string decoded;
+    const char* s = st.c_str();
+    size_t length = strlen(s);
```

Comment:
Don't needs use strlen, and the  null character check can be loop condition with length check.

Reply:
That code is came from the [webview](https://github.com/webview/webview) project, so I left it as is.  It may have something to do with the fact that `std::string` can hold null values, so `std::string::size()` will return an incorrect size,  As far as I know, a null isn't valid values to encode in a URL.  It's up to you, I can change it so it uses `std::string::size()`.

Reply:
More safe code:
```cpp
const char* s = st.c_str();
size_t length = st.length();
for (unsigned int i = 0; i < length; ++i)
{
   if(![i]) break;
   if (s[i] == '%')
    {
        decoded.push_back(hex2char(s + i + 1));
        i = i + 2;
    }
    else if (s[i] == '+')
    {
        decoded.push_back(' ');
    }
    else
    {
        decoded.push_back(s[i]);
    }
}
```

Reply:
Yes, that would stop the decoding if a null value is hit, since it shouldn't be there in the encoded string.  I'll update it now.